lib/Sema/BuilderTransform.cpp

```diff
@@ -367,10 +367,33 @@ class BuilderClosureVisitor
     if (!cs || hadError)
       return nullptr;
 
-    // Call Builder.buildBlock(... args ...)
-    auto call = buildCallIfWanted(braceStmt->getStartLoc(),
-                                  ctx.Id_buildBlock, expressions,
-                                  /*argLabels=*/{ });
+    Expr *call = nullptr;
+    // If the builder supports `buildBlock(combining:into:)`, use this to
+    // combine subexpressions pairwise.
+    if (!expressions.empty() &&
+        builderSupports(ctx.Id_buildBlock, {ctx.Id_combining, ctx.Id_into})) {
+      // Builder.buildBlock(
+      //   combining: arg_n, into: Builder.buildBlock(
+      //     ...
+      //       combining: arg_2, into: Builder.buildBlock(
+      //         combining: arg_1, into: Builder.buildBlock(arg_0))...
+      call = buildCallIfWanted(
+          braceStmt->getStartLoc(), ctx.Id_buildBlock, {expressions.front()},
+          /*argLabels=*/{});
```

Comment:
It might be cleaner to say that we call `buildBlock()` initially to get the first `call`, then the expressions (if there are any) are combined with `buildBlock(_combining:into:)`. That way, we don't need the `expressions.empty()` check in the `if` statement.

Reply:
If we called the nullary `buildBlock()` as a base case, the result builder would have to to support an empty block. Should the developer be allowed to define a DSL that doesn't support empty blocks?